### ISSUE TITLE
Wrong result after shifting unsigned

### ISSUE THREAD

**Issue by bernardofca** [jpoirier/picoc#12]
The following code:

unsigned crc;
crc = 0x80000000;
crc >>= 1;
printf("crc = %X\n", crc);

produces the following output:

crc = C0000000

Which is wrong (should be 4000000)

**Comment by bernardofca** [comment]
The original code (before fork) has the same issue.

**Comment by mwm** [comment]
It's working right for me as of 6ed4a.

**Comment by jpoirier** [comment]
Just confirming what @mwm stated, that it's working for me as well on Fedora 26.

**Comment by bernardofca** [comment]
Thanks to both for spending the time looking at this. I was looking at this project because I am looking at adding a small scripting engine to a an embedded C application. The embedded application will compile with GCC but I like to prototype things with Visual Studio on Windows. 

The code of the right shift assignment is here:

void ExpressionInfixOperator(struct ParseState *Parser,
    struct ExpressionStack **StackTop, enum LexToken Op,
    struct Value *BottomValue, struct Value *TopValue)
{
../..
       long TopInt = ExpressionCoerceInteger(TopValue);                 <== TopInt is 0x00000001
       long BottomInt = ExpressionCoerceInteger(BottomValue);    <== BottomInt is 0x80000000
../..
        case TokenShiftRightAssign:
            ResultInt = ExpressionAssignInt(Parser, BottomValue,
                BottomInt>>TopInt, false);
            break;

 BottomInt>>TopInt <== we have a long (signed) shifted right and a sign extension added to the result.

After further search it appears that this exact behavior is compiler specific. So the C code listed above is compiler specific (it seems that GCC does sign extension like MSFT) but the Basic code is using unsigned variable so it should not fail. I am still puzzled as how it works for you. 

![image](https://user-images.githubusercontent.com/30217698/29285022-d776f7d2-80e1-11e7-955f-f649576150e9.png)

**Comment by mwm** [comment]
Your test doesn't sign extend on our GCC because we're on 64-bit platforms, so 0x80000000 isn't a signed value. If I try the test with 0x8000000000000000 instead, it fails as you describe.

**Comment by bernardofca** [comment]
Good point. Thank you for the analysis. I tried to compile for 64-bit on MSFT but a _long_ is still 4 bytes so it did not help!

**Comment by mwm** [comment]
Sorry for leaving this here, but github turned off message.

I've refactored the picoc code to make supporting multiple platforms easier in order to check it out for an embedded system. You interested in a copy of that?

**Comment by jpoirier** [comment]
Sure but I've been putting in some really long hours at work lately and may not get a chance to look at it until the weekend. BTW, have either of you looked at any of the forks from the original github repo; they may be more up to date and certainly closer to the original since I ripped a lot of the cross platform stuff out IIRC.

**Comment by mwm** [comment]
I've put my version at https://github.com/mwm/picoc.git. I hadn't really planned on creating a PR, so everything is on master. Sorry about that. It did merge with the develop branch cleanly, though.

I checked out a couple of the more recent forks, and this had more work in it. I tried merging one with the STM32 port in it, but that repository wasn't cloned from the original, but started with a code commit so it had no common ancestors.
